doc/pwman.1
.TH PWMAN 1 "March 26, 2002"
.nh
.SH NAME
pwman \- curses based password storage program
.SH SYNOPSIS
.B pwman
[ \fB--help\fP | \fB--version\fP | \fB--gpg-path <path>\fP | \fB--gpg-id\fP <id> | \fB--file\fP <file> | \fB--passphrase-timeout <time in minutes> ] 
.SH DESCRIPTION
This manual page documents briefly the
.B pwman
program.
.PP
.B pwman
is a curses based password storage application. It uses GnuPG to encrypt and decrypt the password file.
The curses user interface was inspired by Jaako Heinonen's abook. 
.SH OPTIONS
.TP
\fB\-\-help\fP
Show usage.
.TP
\fB\-\-version\fP
Show version information.
.TP
\fB\-\-gpg-path\fP <path>
Specify path to GnuPG Executable.
.TP
\fB\-\-gpg-id\fP <id>
Specify GnuPG key to use while encrypting.

This can either be a key ID, or an email address.
.TP
\fB\-\-file\fP <file>
File to read passwords from.
.TP
\fB\-\-passphrase-timeout\fP <time in minutes>
Time before passphrase times out.
.TP
Press '\fB?\fP' during use to get a list of commands.
.SH SEE ALSO
.BR gpg (1),
.br
.SH AUTHOR
.B pwman 
was written by Ivan Kelly <ivan@ivankelly.net>


ChangeLog
0.3.5   Nick Burch <gagravarr@users.sourceforge.net>
    * gnupg.c: add support for encrypting data to multiple recipients
    * gnupg.c/actions.c/pwlist.c: make the export of passwords and lists
      handle multiple recipients

0.3.4   Nick Burch <gagravarr@users.sourceforge.net>
    * *.c: implement read only patch from Jon
    * pwlist.c: XML escape data, so strings with a & in them don't break

0.3.3   Ivan Kelly  <ivan@ivankelly.net>
    * *.c: renamed a few functions to make it clearer where they're from
    * uilist.c: fix for cursor going off the screen

0.3.1	Ivan Kelly	<ivan@ivankelly.net>
	* pwman.1: updated man page
	* gnupg.c: large rewrite, switched back from gpgme to using gpg with pipes
	* pwlist.c/actions.c/uilist.c: added sublist functionallity
	* pwlist.c: new format due to addition of sublists
	* convert_pwdb.c: separate executable for converting old database
	* pwdb2csv.c: separate executable to convert a database to a csv file
	* misc.c: added file misc.c and move some misc functions to it
	* *.c: other small changes
	
0.2.1	Ivan Kelly 	<ivan@ivankelly.net>
	* filter.c: Bugfixes on filter
	* launch.c: Fixed launch command with filters
	* pwlist.c: Added a file write after each add so that data isn't lost if pwman
	  isn't shut down right

0.2.0	Ivan Kelly	<ivan@ivankelly.net>
	* filter.c: Added search/filtering function
	* gnupg.c: Switched from using gpg to gpgme
	* pwgen.c: Added password generator
	* pwlist.c: Rewrote most of password file I/O
	* import_export.c: Added Import/Export functions
	* uilist.c: Got rid of incessant blinking/flashing on scroll
	* ui.c/pwman.c/gnupg.c: Shadowed passphrase and timeout improvements

0.1.2	Ivan Kelly 	<ivan@ivankelly.net>
	* options.c/pwlist.c: Switched from libxml to libxml2
	* gnupg.c: Added a passphrase timeout option
	* configure.in: Changed configure.in so that it now builds on openbsd
	* pwman.c Added lock files on the password files
	* *: Now Building on MacOSX10.1, FreeBSD 4.5 & Solaris 8 also

0.1.1	Ivan Kelly	<ivan@ivankelly.net>
	* *.c: Few bug fixes and error handling. Code clean up.

0.1.0 	Ivan Kelly	<ivan@ivankelly.net>
	* pwman: Initial release


src/pwman.c
/*
 *  PWMan - password management application
 *
 *  Copyright (C) 2002  Ivan Kelly <ivan@ivankelly.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "pwman.h"
#include <signal.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdarg.h>

static void pwman_parse_command_line(int argc, char **argv);
static void pwman_show_usage();
static void pwman_show_version();
static void pwman_quit();


static int
pwman_check_lock_file()
{
	char fn[STRING_LONG];
	FILE *fp;
	
	snprintf(fn, STRING_LONG, "%s.lock", options->password_file);
	if(access(fn, F_OK) == 0){
		return 1;
	} else {
		return 0;
	}
}

static int
pwman_create_lock_file()
{
	char fn[STRING_LONG];
		
	snprintf(fn, STRING_LONG, "%s.lock", options->password_file);
	creat(fn, S_IRWXU);
}

static int
pwman_delete_lock_file()
{
	char fn[STRING_LONG];
	
	snprintf(fn, STRING_LONG, "%s.lock", options->password_file);
	unlink(fn);
}

static void
pwman_init(int argc, char *argv[])
{
	char c;
	signal(SIGKILL, pwman_quit);
	signal(SIGTERM, pwman_quit);

	umask(DEFAULT_UMASK);

	/* get options from .pwmanrc */
	options = options_new();
	if(options_read() == -1){
		options_get();
	}

	/* parse command line options */
	pwman_parse_command_line(argc, argv);

	/* check to see if another instance of pwman is open */
	if(!options->readonly && pwman_check_lock_file()){
		fprintf(stderr, "It seems %s is already opened by an instance of pwman\n",
				options->password_file);
		fprintf(stderr, "Two instances of pwman should not be open the same file at the same time\n");
		fprintf(stderr, "If you are sure pwman is not already open you can delete the file.\n");
		fprintf(stderr,"Alternatively, you can open the file readonly by answering 'r'\n");
		fprintf(stderr,"Delete file %s.lock? [y/n/r]\n",
				options->password_file);
		c = getchar();
		switch (tolower(c)) {
			case 'y':
				pwman_delete_lock_file();
				break;
			case 'r':
				options->readonly = TRUE;
				break;
			default:
				exit(-1);
		}
	}
	
	if( ui_init() ){
		exit(1);
	}

	ui_refresh_windows();

	/* get pw database */
	pwlist_init();
	if(pwlist_read_file() != 0){
		pwlist = pwlist_new("Main");
		current_pw_sublist = pwlist;
	}
	if (!options->readonly){
		pwman_create_lock_file();
	}

	ui_refresh_windows();
}

static void
pwman_quit()
{
	pwlist_write_file();
	pwlist_free_all();
	pwman_delete_lock_file();
	
	ui_end();
	options_write();
	
	exit(0);
}

int
main(int argc, char *argv[])
{
	pwman_init(argc, argv);

	ui_run();

	pwman_quit();
	return 0;
}

static void
pwman_parse_command_line(int argc, char **argv)
{
	int i;

	for(i = 1; i < argc; i++){
		if( !strcmp(argv[i], "--help") || !strcmp(argv[i], "-h") ){
			pwman_show_usage(argv[0]);
			exit(1);
		} else if( !strcmp(argv[i], "--version") || !strcmp(argv[i], "-v") ){
			pwman_show_version();
			exit(1);
		} else if( !strcmp(argv[i], "--gpg-path") ){
			write_options = FALSE;
			strncpy(options->gpg_path, argv[i + 1], STRING_LONG);
			i++;
		}else if( !strcmp(argv[i], "--gpg-id") ){
			write_options = FALSE;
			strncpy(options->gpg_id, argv[i + 1], STRING_LONG);
			i++;
		} else if( !strcmp(argv[i], "--file") || !strcmp(argv[i], "-f") ){
			write_options = FALSE;
			strncpy(options->password_file, argv[i + 1], STRING_LONG);
			i++;
		} else if( !strcmp(argv[i], "--passphrase-timeout") || !strcmp(argv[i], "-t") ){
			write_options = FALSE;
			options->passphrase_timeout = atoi(argv[i + 1]);
			i++;
		} else if ( !strcmp(argv[i], "--readonly") || !strcmp(argv[i], "-r") ){
			write_options = FALSE;
			options->readonly = TRUE;
		} else {
			printf("option %s not recognised\n", argv[i]);
			printf("try %s --help for more info\n", argv[0]);
			exit(1);
		}
	}
}

static void
pwman_show_version()
{
	puts(PACKAGE " v " VERSION);
	puts("Written by Ivan Kelly <ivan@ivankelly.net>\n");
	puts("Copyright (C) 2002 Ivan Kelly");
	puts("This program is free software; you can redistribute it and/or modify");
	puts("it under the terms of the GNU General Public License as published by");
	puts("the Free Software Foundation; either version 2 of the License, or");
	puts("(at your option) any later version.\n");

	puts("This program is distributed in the hope that it will be useful,");
	puts("but WITHOUT ANY WARRANTY; without even the implied warranty of");
	puts("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the");
	puts("GNU General Public License for more details.\n");

	puts("You should have received a copy of the GNU General Public License");
	puts("along with this program; if not, write to the Free Software");
	puts("Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111, USA.\n");
}

static void
pwman_show_usage(char *argv_0)
{
	printf("Usage: %s [OPTIONS]...\n", argv_0);
	puts("Store you passwords securely using public key encryption\n");
	puts("  --help                 show usage");
	puts("  --version              display version information");
	puts("  --gpg-path <path>      Path to GnuPG executable");
	puts("  --gpg-id <id>          GnuPG ID to use");
	puts("  --file <file>          file to read passwords from");
	puts("  --passphrase-timeout <mins>    time before app forgets passphrase(in minutes)");
	puts("  --readonly             open the database readonly\n\n");
	puts("Report bugs to <ivan@ivankelly.net>");
}
